zookeeper-specifications/zab-1.0/ZabWithFLE.tla

```diff
@@ -0,0 +1,1020 @@
+                       ackIndex, currentCounter, sendCounter, committedIndex, tempVarsZ, followerVarsZ, verifyVarsZ, electionMsgs, idTable>>
+
+LeaderSync2(i) ==
+        /\ state[i] = LEADING
+        /\ zabState[i] = SYNCHRONIZATION
+        /\ ackldRecv[i] \in Quorums
+        /\ commitIndex'    = [commitIndex    EXCEPT ![i] = Len(history[i])]
+        /\ committedIndex' = [committedIndex EXCEPT ![i] = Len(history[i])]
+        /\ zabState'       = [zabState       EXCEPT ![i] = BROADCAST]
+        /\ currentCounter' = [currentCounter EXCEPT ![i] = 0]
+        /\ sendCounter'    = [sendCounter    EXCEPT ![i] = 0]
+        /\ ackldRecv'      = [ackldRecv      EXCEPT ![i] = ackldRecv[i] \union {NullPoint}]
+        /\ BroadcastUPTODATE(i, [mtype   |-> UPTODATE,
+                                 mepoch  |-> acceptedEpoch[i],
+                                 mcommit |-> Len(history[i])]) \* In actual UPTODATE doesn't carry this info
+        /\ UNCHANGED <<state, currentEpoch, lastZxid, acceptedEpoch, history, electionVarsZ, leadingVoteSet, learners, cepochRecv, ackeRecv, forwarding, ackIndex, 
+                       committedCounter, tempVarsZ, followerVarsZ, verifyVarsZ, electionMsgs, idTable>>
+
+FollowerHandleUPTODATE(i, j) ==
+        /\ state[i] = FOLLOWING
+        /\ msgs[j][i] /= << >>
+        /\ msgs[j][i][1].mtype = UPTODATE
+        /\ LET msg == msgs[j][i][1]
+               infoOk == /\ leaderAddr[i] = j
+                         /\ acceptedEpoch[i] = msg.mepoch
+               correct == /\ infoOk
+                          /\ zabState[i] = SYNCHRONIZATION
+                          /\ currentEpoch[i] = msg.mepoch
+           IN /\ infoOk
+              /\ \/ /\ correct
+                    /\ commitIndex' = [commitIndex EXCEPT ![i] = Maximum({commitIndex[i], msg.mcommit})] \* May have received COMMIT and have a bigger commit index.
+                    /\ zabState'    = [zabState    EXCEPT ![i] = BROADCAST]
+                    /\ UNCHANGED inherentViolated
+                 \/ /\ ~correct
+                    /\ PrintT("Exception: Condition correct is false in FollowerHandleUPTODATE(" \o ToString(i) \o ", " \o ToString(j) \o ").")
+                    /\ inherentViolated' = TRUE
+                    /\ UNCHANGED <<commitIndex, zabState>>
+        /\ Discard(j, i)
+        /\ UNCHANGED <<state, currentEpoch, lastZxid, acceptedEpoch, history, electionVarsZ, leaderVarsZ, tempVarsZ, followerVarsZ,
+                       proposalMsgsLog, epochLeader, electionMsgs, idTable>>
+
+-----------------------------------------------------------------------------
+(* In phase l31, leader receives client request and broadcasts PROPOSAL.
+   Note: In production, any server in traffic can receive requests and forward it to leader if necessary.
+   We choose to let leader be the sole one who can receive requests, to simplify spec and keep correctness at the same time. *)
+ClientRequest(i, v) ==
```

Comment:
Nothing is stopping the same value being sent over and over again. Usually use a set for values when you want to limit one client request per set member. So with {v1, v2}, no history could be longer than 2 and never have duplicates.
I think you need to add an enabling condition that checks that the value does not already exist in a history.